### ISSUE TITLE
A way to avoid certain structs being passed between async -> await statements

### ISSUE THREAD

**Issue by Havunen** [rust-lang/rust-clippy#15777]
### What it does

One example for the use case:

Lets imagine following code:

```rs
pub async fn add_new_organization() {
   let db_conn = aquire_db_connection();
   
   do_something_async().await;

   insert_database_data(&mut db_conn);
}
```

While the code works as expected in the current state, I would like to clippy throw error because the above code holds to database connection and the do_something_async(); can take for a long time to finish which can cause database connection exhaust in high traffic scenario.

The following version should not throw warning:

```rs
pub async fn add_new_organization() {  
   do_something_async().await;
   let db_conn = aquire_db_connection();

   insert_database_data(&mut db_conn);
}
```

or if db_conn is dropped before the `await` statement. Maybe there could be some way to decorate such structs which should not exist between await statements, so users can select which structs to track.

### Advantage

- Removes the need to manually investigate certain structs are not kept in memory between `async` - `await`

### Example

```rust
<code>
```

Could be written as:

```rust
<code>
```

**Comment by y21** [comment]
You can use the [`await_holding_invalid_type`](https://rust-lang.github.io/rust-clippy/master/index.html#await_holding_invalid_type) lint for this.

**Comment by Havunen** [comment]
Oh perfect! Thanks @y21 !
